forc-plugins/forc-doc/src/render/util/format/docstring.rs

```diff
@@ -0,0 +1,91 @@
+        options.parse.smart = true;
+        options.parse.default_info_string = Some(SWAY_FILEINE.into());
+        markdown_to_html(&format_docs(&docs), &options)
+    }
+    fn to_raw_string(&self) -> String {
+        let attributes = self.get(&AttributeKind::DocComment);
+        let mut docs = String::new();
+
+        if let Some(vec_attrs) = attributes {
+            for arg in vec_attrs.iter().flat_map(|attribute| &attribute.args) {
+                writeln!(docs, "{}", arg.name.as_str()).expect(
+                    "problem appending `arg.name.as_str()` to `docs` with `writeln` macro.",
+                );
+            }
+        }
+        docs
+    }
+}
+
+/// Create a docstring preview from raw html attributes.
+///
+/// Returns `None` if there are no attributes.
+pub(crate) fn create_preview(raw_attributes: Option<String>) -> Option<String> {
+    raw_attributes.as_ref().map(|description| {
+        let preview = split_at_markdown_header(description);
+        if preview.chars().count() > MAX_PREVIEW_CHARS && preview.contains(CLOSING_PARAGRAPH_TAG) {
+            let closing_tag_index = preview
+                .find(CLOSING_PARAGRAPH_TAG)
+                .expect("closing tag out of range");
+            // We add 1 here to get the index of the char after the closing tag.
+            // This ensures we retain the closing tag and don't break the html.
+            let (preview, _) =
+                preview.split_at(closing_tag_index + CLOSING_PARAGRAPH_TAG.len() + 1);
+            if preview.chars().count() > MAX_PREVIEW_CHARS && preview.contains(NEWLINE_CHAR) {
+                let newline_index = preview
+                    .find(NEWLINE_CHAR)
+                    .expect("new line char out of range");
+                preview.split_at(newline_index).0.to_string()
+            } else {
+                preview.to_string()
+            }
+        } else {
+            preview.to_string()
+        }
+    })
+}
+
+/// Checks if some raw html (rendered from markdown) contains a header.
+/// If it does, it splits at the header and returns the slice that preceeded it.
+pub(crate) fn split_at_markdown_header(raw_html: &str) -> &str {
+    if raw_html.contains(H1) {
+        let v: Vec<_> = raw_html.split(H1).collect();
+        v.first().expect("expected a non-empty str")
+    } else if raw_html.contains(H2) {
+        let v: Vec<_> = raw_html.split(H2).collect();
+        v.first().expect("expected a non-empty str")
+    } else if raw_html.contains(H3) {
+        let v: Vec<_> = raw_html.split(H3).collect();
+        v.first().expect("expected a non-empty str")
+    } else if raw_html.contains(H4) {
+        let v: Vec<_> = raw_html.split(H4).collect();
+        v.first().expect("expected a non-empty str")
+    } else if raw_html.contains(H5) {
+        let v: Vec<_> = raw_html.split(H5).collect();
+        v.first().expect("expected a non-empty str")
+    } else {
+        raw_html
+    }
```

Comment:
hm, I have a feeling that there's probably some slick way of refining this into a few lines

Reply:
I could put them in a vec, then loop through the vec and perform the action that's being repeated here. That would reduce the amount of lines for sure. I'm honestly not sure why I did it this way initially, I think I was just trying to rush past this part and finish the main portion of the PR this came from.